### PR TITLE
[FEAT] Get list of batches

### PR DESCRIPTION
**Pull request by quersa**
### Description

This MR exposes batches list end point to the sdk with filtering.

#### Related PRs in other projects (PASQAL developers only)

Closely related to "Get list of jobs":  https://github.com/pasqal-io/pasqal-cloud/pull/121

#### Additional merge criteria

None

#### Breaking changes

None

### Checklist

- [x] The title of the PR follows the right format: [{Label}] {Short Message}. Label examples: IMPROVEMENT, FIX, REFACTORING... Short message is about what your PR changes.

#### Versioning (PASQAL developers only)

- [x] Update the version of pasqal-cloud in `_version.py` following the changes in your PR and by using [semantic versioning](https://semver.org/).

#### Documentation

- [x] Update CHANGELOG.md with a description explaining briefly the changes to the users.

#### Tests

- [x] Unit tests have been added or adjusted.
- [x] Tests were run locally.

#### Internal tests pipeline (PASQAL developers only)
- [x] Update and run the internal tests while targeting the branch of this PR.
 If your PR hasn't changed any functionality, it still needs to be validated against internal tests.

#### After updating the version (PASQAL developers only)

- [ ] Open a PR on the internal tests that updates the version used for the pasqal-cloud backward compatibility tests.

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -8,6 +8,7 @@ All notable changes to this project will be documented in this file.
 
 - Introduced a new filtering system using classes for more flexible and robust filtering options
 - Added 'get_jobs' method, a feature to retrieve a list of jobs based on 'JobFilters'
+- Added 'get_batches' method, a feature to retrieve a list of batches based on 'BatchFilters'
```
Comment (Mildophin) [inline CHANGELOG.md:11]: So in this case you would need to add a new section with a Added subtitle and a newer version (add +1 to the patch version of the latest released version).

File: pasqal_cloud/_version.py [hunk 2]
```diff
@@ -13,4 +13,4 @@
 # limitations under the License.
 
 
-__version__ = "0.11.1"
+__version__ = "0.11.2"
```
Comment (Mildophin) [inline pasqal_cloud/_version.py:16]: As I just released it, rn it would be 0.11.3 instead but there is https://github.com/pasqal-io/pasqal-cloud/pull/131 open so if they merge it before you, you will have to update it to 0.11.4 (don't forget to update the changelog version as well if it happens)

File: pasqal_cloud/__init__.py [hunk 3]
```diff
@@ -229,6 +230,81 @@ def get_batch(self, id: str, fetch_results: bool = False) -> Batch:
...
+        Returns:
+            PaginatedResponse: A class instance with two fields:
+                - total: An integer representing the total number of batches
+                         matching the filters.
+                - results: A list of batches matching the filters and pagination
+                           parameters provided.
+
+        Raises:
+            BatchFetchingError: If fetching batches call failed.
+            ValueError: If `filters` is not an instance of JobFilters
```
Comment (Mildophin) [inline pasqal_cloud/__init__.py:271]: typo: BatchFilters

File: pasqal_cloud/utils/models.py [hunk 4]
```diff
@@ -76,6 +76,74 @@ def status_enum_to_string(self, job_statuses: List[JobStatus]) -> List[str]:
...
+
+    When using several filters at the same time, the API will return elements who pass
+    all filters at the same time:
+        - If the filter value is a single element, the API will return batches whose
+          attribute matches the provided value.
+        - If the filter value is a list, the API will return batches whose value for
+          this attribute is contained in that list.
+
+    Attributes:
+        id: Filter by job IDs.
```
Comment (Mildophin) [inline pasqal_cloud/utils/models.py:93]: Batch*

File: pasqal_cloud/utils/models.py [hunk 5]
```diff
@@ -76,6 +76,74 @@ def status_enum_to_string(self, job_statuses: List[JobStatus]) -> List[str]:
...
+        project_id: Filter by project IDs.
+        user_id: Filter by user IDs.
+        device_type: Filter by device type.
+        status: Filter by job statuses.
```
Comment (Mildophin) [inline pasqal_cloud/utils/models.py:97]: Batch statuses*

File: pasqal_cloud/utils/models.py [hunk 6]
```diff
@@ -76,6 +76,74 @@ def status_enum_to_string(self, job_statuses: List[JobStatus]) -> List[str]:
...
+        min_runs: Minimum number of runs.
+        max_runs: Maximum number of runs.
```
Comment (Mildophin) [inline pasqal_cloud/utils/models.py:99]: runs should be removed from here

File: tests/test_batch.py [hunk 7]
```diff
@@ -607,3 +614,190 @@ def test_rebatch_raises_value_error_on_invalid_filters(self):
...
+        ],
+    )
+    def test_get_batches_with_filters_success(
+        self,
+        mock_request: Any,
+        filters: Union[BatchFilters, None],
+    ):
+        """
+        As a user using the SDK with proper credentials,
+        I can get a list of jobs with specific filters.
```
Comment (Mildophin) [inline tests/test_batch.py:680]: batches*, below as well

File: tests/test_batch.py [hunk 8]
```diff
@@ -607,3 +614,190 @@ def test_rebatch_raises_value_error_on_invalid_filters(self):
...
+            PaginationParams(offset=1000, limit=100),
+        ],
+    )
+    def test_get_batches_with_pagination_success(
+        self, mock_request: Any, pagination: PaginationParams
+    ):
+        """
+        As a user using the SDK with proper credentials,
+        I can get a list of jobs with pagination parameters.
+        The resulting request will retrieve the jobs that match
```
Comment (Mildophin) [inline tests/test_batch.py:718]: batches*

File: tests/test_batch.py [hunk 9]
```diff
@@ -607,3 +614,190 @@ def test_rebatch_raises_value_error_on_invalid_filters(self):
...
+            == f"{self.sdk._client.endpoints.core}/api/v2/batches{query_params}"
+        )
+
+    def test_get_batches_sdk_error(
+        self, mock_request_exception: Generator[Any, Any, None]
+    ):
+        """
+        As a user using the SDK with proper credentials,
+        if my request for getting jobs returns a status code different from 200,
+        I am faced with the JobFetchingError.
```
Comment (Mildophin) [inline tests/test_batch.py:748]: BatchFetchingError*

File: tests/test_batch.py [hunk 10]
```diff
@@ -607,3 +614,190 @@ def test_rebatch_raises_value_error_on_invalid_filters(self):
...
+            # List of user_ids as strings
+            BatchFilters(user_id=["1", "2"]),
+            # Single UUID for batch_id
+            BatchFilters(id=UUID(int=0x2)),
+            # Single status
+            BatchFilters(status=BatchStatus.DONE),
+            # List of statuses
+            BatchFilters(status=[BatchStatus.DONE, BatchStatus.PENDING]),
+            # Minimum runs
+            BatchFilters(min_runs=10),
```
Comment (Mildophin) [inline tests/test_batch.py:652]: all runs related filters should not exists, please remove them

**Review by Mildophin — CHANGES_REQUESTED** [review]
Lastly, something to mention is as augustin is working on https://github.com/pasqal-io/pasqal-cloud/pull/131, you will probably have to rebase and change all the complete to open (and reverse the boolean value so). I would suggest to rebase your PR on his PR to avoid surprises.

File: pasqal_cloud/__init__.py [hunk 11]
```diff
@@ -230,6 +231,81 @@ def get_batch(self, id: str, fetch_results: bool = False) -> Batch:
...
+        Raises:
+            BatchFetchingError: If fetching batches call failed.
+            ValueError: If `filters` is not an instance of BatchFilters
+                    or if `pagination_params` is not an instance of PaginationParams.
+        """
+
+        if pagination_params is None:
+            pagination_params = PaginationParams()
+        elif not isinstance(pagination_params, PaginationParams):
+            raise ValueError(
```
Comment (MatthieuMoreau0) [inline pasqal_cloud/__init__.py:279]: nit: `TypeError` seems more appropriate

File: pasqal_cloud/__init__.py [hunk 12]
```diff
@@ -230,6 +231,81 @@ def get_batch(self, id: str, fetch_results: bool = False) -> Batch:
...
+                f"Pagination parameters needs to be a PaginationParams instance, "
+                f"not a {type(pagination_params)}"
+            )
+
+        if filters is None:
+            filters = BatchFilters()
+        elif not isinstance(filters, BatchFilters):
+            raise ValueError(
```
Comment (MatthieuMoreau0) [inline pasqal_cloud/__init__.py:287]: same comment

File: pasqal_cloud/__init__.py [hunk 13]
```diff
@@ -230,6 +231,81 @@ def get_batch(self, id: str, fetch_results: bool = False) -> Batch:
...
 
+    def get_batches(
+        self,
+        filters: Optional[BatchFilters] = None,
+        pagination_params: Optional[PaginationParams] = None,
+    ) -> PaginatedResponse:
+        """
+        Retrieve a list of batches matching filters using a pagination system.
+
+        batches are sorted by their creation date in descending order.
```
Comment (MatthieuMoreau0) [inline pasqal_cloud/__init__.py:242]: @Mildophin I just discovered the user cannot chose the ordering in the pagination params? Is that voluntary? Should that be added later?

File: pasqal_cloud/utils/filters.py [hunk 14]
```diff
@@ -4,17 +4,15 @@
...
 from pydantic import BaseModel, Field, field_serializer, field_validator
 
-from pasqal_cloud.utils.constants import JobStatus
+from pasqal_cloud.utils.constants import BatchStatus, JobStatus, QueuePriority
 
 
-class BaseJobFilters(BaseModel):
+class BaseFilters(BaseModel):
     """
     Base class used by Job related filters for shared fields.
```
Comment (MatthieuMoreau0) [inline pasqal_cloud/utils/filters.py:12]: docnit: this doesn't apply only to job anymore so docstring should be updated

**Review by MatthieuMoreau0 — APPROVED** [review]
Looks good! Couple nitpicks only

File: pasqal_cloud/__init__.py (hunk 11 above)
Comment (Mildophin) [reply pasqal_cloud/__init__.py:279]: You are right, I used Valuerror on my previous PRs related to bulk actions, should I update them you think ?

File: pasqal_cloud/__init__.py (hunk 13 above)
Comment (Mildophin) [reply pasqal_cloud/__init__.py:242]: It was voluntary as I didn't understood it as a feature in the scope of the ticket. I think we can add it later yes.

File: pasqal_cloud/utils/constants.py [hunk 15]
```diff
@@ -1,9 +1,40 @@
 from enum import Enum
 
 
+class Status(Enum):
```
Comment (MohamedAminMallek) [inline pasqal_cloud/utils/constants.py:4]: I believe this enum is not being used

**Review by MohamedAminMallek — APPROVED** [review]

File: pasqal_cloud/__init__.py [hunk 16]
```diff
@@ -231,6 +232,81 @@ def get_batch(self, id: str, fetch_results: bool = False) -> Batch:
...
+        >>> Get_batches(filters=BatchFilters(status=BatchStatus.ERROR),
+                     pagination_params=PaginationParams(offset=200))
+        Returns batches 201-300 with an ERROR status.
+
+        Returns:
+            PaginatedResponse: A class instance with two fields:
+                - total: An integer representing the total number of batches
+                         matching the filters.
+                - results: A list of batches matching the filters and pagination
+                           parameters provided.
```
Comment (MatthieuMoreau0) [inline pasqal_cloud/__init__.py:269]: This shouldn't be documented here, there is a strong risk this becomes desynchronized with the PaginatedResponse class. 

We should just clarify this returns a PaginatedResponse instance which includes the results of the API and some pagination information

File: pasqal_cloud/utils/constants.py [hunk 17]
```diff
@@ -1,9 +1,32 @@
 from enum import Enum
 
 
+class BatchStatus(Enum):
+    PENDING = "PENDING"
+    RUNNING = "RUNNING"
+    DONE = "DONE"
+    ERROR = "ERROR"
+    CANCELED = "CANCELED"
```
Comment (MatthieuMoreau0) [inline pasqal_cloud/utils/constants.py:9]: you're missing TIMED_OUT

**Review by MatthieuMoreau0 — APPROVED** [review]

File: CHANGELOG.md [hunk 18]
```diff
@@ -2,20 +2,26 @@
...
 - Deprecation warnings on init of SDK, now each version of pasqal-cloud is supported for 1 year after release.
   If your version is not supported anymore you will get a deprecation warning each time you use the SDK.
 
-## [0.12.3]
-
-_released `2024-10-02`_
+## [0.12.3] - 2024-10-02
 
-### ✨ Added
+### Added
```
Comment (Augustinio) [inline CHANGELOG.md:24]: Why was this modified?

**Review by Augustinio — APPROVED** [review]
Looks good aside from the modified sections in the changelog
